apps/studio/components/interfaces/Organization/BillingSettings/CreditTopUp.tsx

```diff
@@ -0,0 +1,369 @@
+      paymentMethod: '',
+    },
+  })
+
+  const [topUpModalVisible, setTopUpModalVisible] = useState(false)
+  const [paymentConfirmationLoading, setPaymentConfirmationLoading] = useState(false)
+  const captchaLoaded = useIsHCaptchaLoaded()
+  const [captchaToken, setCaptchaToken] = useState<string | null>(null)
+  const [captchaRef, setCaptchaRef] = useState<HCaptcha | null>(null)
+
+  const captchaRefCallback = useCallback((node: any) => {
+    setCaptchaRef(node)
+  }, [])
+
+  const resetCaptcha = () => {
+    setCaptchaToken(null)
+    captchaRef?.resetCaptcha()
+  }
+
+  const initHcaptcha = async () => {
+    if (topUpModalVisible && captchaRef && captchaLoaded) {
+      let token = captchaToken
+
+      try {
+        if (!token) {
+          const captchaResponse = await captchaRef.execute({ async: true })
+          token = captchaResponse?.response ?? null
+          setCaptchaToken(token)
+          return token
+        }
+      } catch (error) {
+        return token
+      }
+
+      return token
+    }
+  }
+
+  useEffect(() => {
+    initHcaptcha()
+  }, [topUpModalVisible, captchaRef, captchaLoaded])
+
+  const [paymentIntentSecret, setPaymentIntentSecret] = useState('')
+  const [paymentIntentConfirmation, setPaymentIntentConfirmation] = useState<PaymentIntentResult>()
+
+  const onSubmit: SubmitHandler<CreditTopUpForm> = async ({ amount, paymentMethod }) => {
+    setPaymentIntentConfirmation(undefined)
+
+    const token = await initHcaptcha()
+
+    await topUpCredits(
+      {
+        slug,
+        amount,
+        payment_method_id: paymentMethod,
+        hcaptchaToken: token,
+      },
+      {
+        onSuccess: (data) => {
+          if (data.status === 'succeeded') {
+            onSuccessfulPayment()
+          } else {
+            setPaymentIntentSecret(data.payment_intent_secret || '')
+          }
+
+          resetCaptcha()
+        },
+      }
+    )
+  }
+
+  const options = useMemo(() => {
+    return {
+      clientSecret: paymentIntentSecret,
+      appearance: { theme: resolvedTheme?.includes('dark') ? 'night' : 'flat', labels: 'floating' },
+    } as any
+  }, [paymentIntentSecret, resolvedTheme])
+
+  const onTopUpDialogVisibilityChange = (visible: boolean) => {
+    setTopUpModalVisible(visible)
+    if (!visible) {
+      setCaptchaRef(null)
+      setPaymentIntentConfirmation(undefined)
+      setPaymentIntentSecret('')
+    }
+  }
+
+  const paymentIntentConfirmed = (paymentIntentConfirmation: PaymentIntentResult) => {
+    // Reset payment intent secret to ensure another attempt works as expected
+    setPaymentIntentSecret('')
+    setPaymentIntentConfirmation(paymentIntentConfirmation)
+
+    if (paymentIntentConfirmation.paymentIntent?.status === 'succeeded') {
+      onSuccessfulPayment()
+    }
+  }
+
+  const onSuccessfulPayment = async () => {
+    onTopUpDialogVisibilityChange(false)
+    await queryClient.invalidateQueries(subscriptionKeys.orgSubscription(slug))
+    toast.success(
```

Comment:
nit: Input could be reset to 100 after a successful top up? Else if i open the modal again, my previous value is persisted

Reply:
Think it's fine, pretty unlikely to do two top ups in a row and if so the amount can stay the same